fix(material/datepicker): avoid rerender when min/maxDate changes to different time on the same day

zarend opened the pull request:
Avoid re-rendering <mat-calendar/> when the [minDate] or [maxDate]
Input change to a different time on the same day.

In `ngOnChanges`, do not call `init` if the previous and current value for
minDate/maxDate are on the same day. This makes #24384 a non-breaking
code change.

Fixes #24435

wagnermaciel commented on src/material/datepicker/calendar.ts:
```diff
@@ -255,19 +256,43 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
-    this._maxDate = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
+    this._oldMaxDate = this._maxDate;
+
+    const validated = this._dateAdapter.getValidDateOrNull(this._dateAdapter.deserialize(value));
```
Is it possible to move this logic into a separate function? We seem to be repeating this in minDate's setter

wagnermaciel commented on src/material/datepicker/calendar.ts:
```diff
@@ -393,7 +418,19 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
-    const change = changes['minDate'] || changes['maxDate'] || changes['dateFilter'];
+    const minDateChange: SimpleChange | undefined = !this._dateAdapter.sameDate(
+      this._minDate,
+      this._oldMinDate,
```
Do _oldMinDate and _oldMaxDate exist for the sole purpose of comparing them here? If so, SimpleChange has a reference to the previous value you can use - https://angular.io/api/core/SimpleChange

zarend replied:
done ✔️

zarend commented:
@wagnermaciel @crisbeto this PR is ready for your eyes again 👀 . I've addressed the PR comments and it is no longer a draft. This is the most readable solution I could come up with, curious to know what you think.

andrewseguin commented on src/material/datepicker/calendar.ts:
```diff
@@ -393,7 +393,27 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
-    const change = changes['minDate'] || changes['maxDate'] || changes['dateFilter'];
+    let change = undefined;
+
+    for (let key in changes) {
```
The use of the for loop that breaks when it finds a change is a little hard to follow. Would prefer to see more straightforward logic here, e.g. something more like

```
const minDateChanged = !this._dateAdapter.sameDate(changes['minDate'].previousValue, changes['minDate'].currentValue) && !changes['minDate'].firstChange;
const minDateChanged = !this._dateAdapter.sameDate(changes['maxDate'].previousValue, changes['maxDate'].currentValue) && !changes['maxDate'].firstChange;
const dateFilterChanged = changes['dateFilter'] && !changes['dateFilter'].firstChange

if (minDateChanged || maxDateChanged || dateFilterChanged) {
  ...
}

```

zarend replied:
Okay, let's not use the for-loop approach.
